[Feature] Support Dify image uploads with data URL normalizatio

springtiming opened the pull request:
## Summary
- Enable Dify adapter to collect and upload images from chat messages, normalizing data URLs and passing upload_file_id in requests.
- Keep additional_kwargs untouched (no auto-injected images) while preserving multimodal metadata via chatluna_multimodal.


## Testing
- Manual: send a message with an image → upload succeeds, Dify workflow receives the file, response returns normally.

springtiming committed:
fix(processDynamicImport): exclude declaration files from dynamic import processing

- Update file filtering logic to skip TypeScript declaration files (.d.ts) when retrieving files from directories.
- Ensure only relevant TypeScript files are processed for dynamic imports, improving overall functionality and reducing potential errors.

springtiming committed:
feat: enhance conversation ID handling and logging across components

- Introduce robust conversation ID resolution in DifyRequester, allowing fallback from parameters and variables.
- Add debug logging to track conversation ID flow in DifyRequester and ChatLunaChatChain.
- Ensure conversation ID is preserved in ChatLunaChatModel and included in requests for downstream processing.
- Improve logging in ChatInterface and ChatInterfaceWrapper to verify conversation ID integrity.
- Update ModelInfo capabilities in DifyClient to include ToolCall support.

springtiming committed:
feat: enhance conversation ID handling and platform initialization

- Add platform loading wait logic in _initModel
- Add model value availability check with timeout
- Improve debug logging for model initialization
- Enhance conversation ID handling in Dify adapter

springtiming committed:
feat: add Dify local_file uploads

springtiming committed:
chore: lint fixes and type declarations

springtiming committed:
chore: remove generated d.ts files

dingyi222666 committed:
refactor(adapter-dify): extract file handling utilities and clean up debug logs

- Extract file resolution, MIME type mapping, and multimodal serialization logic into separate utils.ts module
- Remove undici dependency from package.json (moved to shared dependencies)
- Clean up all debug/trace logging statements across core and adapter-dify
- Add platform variable to system prompt variables in core/utils/string.ts
- Pass all custom variables from chain to Dify with chatluna_ prefix
- Simplify conversation ID handling by removing redundant checks and temporary workarounds
- Improve code organization and maintainability by separating concerns

dingyi222666 commented on packages/adapter-dify/src/requester.ts:
```diff
@@ -244,6 +334,629 @@
+            this.logger.warn(`Failed to read file from ${filePath}`, error)
+            return null
+        }
+    }
```
do nothing

dingyi222666 committed:
refactor(core): simplify conversation ID resolution in ChatLunaChatModel

Replace multiple fallback checks for conversation ID from configurable options with direct access to variables_hide.built.conversationId for cleaner and more maintainable code.

dingyi222666 commented:
如果可以了话，在此 issue 下回复。**不要在群聊里艾特我**，我不一定看得到。在 issue 底下交流可以保证互相都能看到消息，因为是有抄送为邮件。

dingyi222666 commented on packages/core/src/llm-core/platform/model.ts:
```diff
@@ -166,6 +166,14 @@ export class ChatLunaChatModel extends BaseChatModel<ChatLunaModelCallOptions> {
             maxTokenLimit = this.getModelMaxContextSize()
         }
 
+        // Ensure conversation id is preserved even if LangChain options did not
+        // include it directly. It may be passed through `configurable` when the
+        // executor is invoked.
+        let id = options?.id ?? this._options.id
+        if (!id) {
+            id = options?.variables_hide['built']['conversationId']
+        }
+
```
此处为故意的，方便后续调试诊断。

dingyi222666 commented:
确认可以了我才合并，，，你不测试的话我不知道能不能用

springtiming committed:
fix: improve dify model handling and chat model detection

- add undici to Dify adapter deps and expose getModels override
- clean refreshModels ModelInfo shape and harden non-chat model error
- allow ChatLunaChatModel detection via constructor name to handle module isolation

dingyi222666 commented on packages/core/src/llm-core/chat/app.ts:
```diff
@@ -452,8 +452,16 @@ export class ChatInterface {
-            return [llmModel, llmInfo]
+        // 检查 llmModel.value 是否是 ChatLunaChatModel
+        // 使用 constructor.name 作为后备，因为模块隔离可能导致 instanceof 检查失败
+        if (llmModel.value != null) {
```
这里为啥需要后备检测是否为 `ChatLunaChatModel`，给出原因。dify adapter 默认返回不都是 `ChatLunaChatModel` 吗

springtiming replied:
我在测试时发现就是因为模块隔离虽然可以传入正确的类定义但是无法通过检查会导致303

dingyi222666 commented on packages/adapter-dify/package.json:
```diff
@@ -66,11 +66,12 @@
         "atsc": "^2.1.0",
-        "koishi": "^4.18.9"
+        "koishi": "^4.18.9",
+        "undici": "^6.21.3"
```
undici 是必要的吗？全部使用 `plugin.fetch` 不行吗？你是使用 pnpm 还是啥，没法用么

springtiming replied:
实现中使用的Dify 上传接口需要 multipart/form-data，如果不引用无法完成构建

dingyi222666 reviewed (commented):
need review

springtiming commented:
在 adapter-dify 的 devDependencies 中加入了 undici，仅用于本地/的测试构建；
发布产物仍依赖 core 中的实现，不会引入额外运行时依赖。

同时修复了一个会导致 303 的问题。已通过手动测试验证，功能运行正常。

dingyi222666 commented:
会导致 303，应该是你的本地例子多例了。。。。删掉 lock 文件重新 install。。。这里的判断完全不需要

springtiming committed:
refactor(ChatInterface): simplify ChatLunaChatModel detection logic

xiaolan20020118-create commented:
谢谢两位大哥！

dingyi222666 commented:
新电脑弄好了，下周一或者周二合并

dingyi222666 committed:
refactor(adapter-dify): improve history handling and code formatting

- Add history length limiting (130k chars) with truncation logic
- Extract text content from multimodal messages
- Improve null safety for message access
- Format error message for better readability

dingyi222666 commented:
下次提交 pr 之前，记得给自己的 git 和 github 设置 gpg 签名.... 不然我不合并了，我会全部 squash 到一个我签名的 commit 里....
